modules/apps/transfer/keeper/migrations.go

```diff
@@ -0,0 +1,52 @@
+package keeper
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	"github.com/cosmos/ibc-go/v4/modules/apps/transfer/types"
+)
+
+// Migrator is a struct for handling in-place store migrations.
+type Migrator struct {
+	keeper Keeper
+}
+
+// NewMigrator returns a new Migrator.
+func NewMigrator(keeper Keeper) Migrator {
+	return Migrator{keeper: keeper}
+}
+
+// This migration will update the denom
+func (m Migrator) MigrateTraces(ctx sdk.Context) error {
+	var iterErr error
+
+	// list of traces that must replace the old traces in store
+	var newTraces []types.DenomTrace
+	m.keeper.IterateDenomTraces(ctx,
+		func(dt types.DenomTrace) (stop bool) {
+			// check if the new way of splitting FullDenom
+			// into Trace and BaseDenom passes validation and
+			// is the same as the current DenomTrace.
+			// If it isn't then store the new DenomTrace in the list of new traces.
+			newTrace := types.ParseDenomTrace(dt.GetFullDenomPath())
+			err := newTrace.Validate()
+			if err != nil {
+				iterErr = err
```

Comment:
Maybe it should just panic and then have the chain devs export a snapshot and manually address the situation? What do you think?

Reply:
In which case, couldn't we return the error immediately? Instead of continuing to do unnecessary work... I'm assuming that the upgrade handler and `RunMigrations` will [likely bubble up the error](https://github.com/cosmos/cosmos-sdk/blob/main/types/module/module.go#L444-L447) and revert any changes so I think it seems unnecessary to continue parsing traces if we encounter an error for some unforeseen reason!

Reply:
we do stop the parsing and then return the error in the current code, is there another way to return the error that you think would make more sense?

Reply:
Ah sorry, I missed the `return true` underneath here due to the thread of comments!